### PR TITLE
[Feature] Liu yi/surpport edpose

### PR DESCRIPTION
**Pull request by LiuYi-Up**
## Motivation

Support inference pipline for ED-Pose

## Modification

新文件：   configs/body_2d_keypoint/edpose/coco/edpose_coco.md
新文件：   configs/body_2d_keypoint/edpose/coco/edpose_coco.yml
新文件：   configs/body_2d_keypoint/edpose/coco/edpose_res50_coco.py
新文件：   docs/src/papers/algorithms/edpose.md
修改：     mmpose/codecs/__init__.py
新文件：   mmpose/codecs/edpose_label.py
修改：     mmpose/datasets/transforms/__init__.py
修改：     mmpose/datasets/transforms/bottomup_transforms.py
修改：     mmpose/models/data_preprocessors/__init__.py
修改：     mmpose/models/data_preprocessors/data_preprocessor.py
修改：     mmpose/models/heads/__init__.py
新文件：   mmpose/models/heads/transformer_heads/__init__.py
新文件：   mmpose/models/heads/transformer_heads/base_transformer_head.py
新文件：   mmpose/models/heads/transformer_heads/edpose_head.py
新文件：   mmpose/models/heads/transformer_heads/transformers/__init__.py
新文件：   mmpose/models/heads/transformer_heads/transformers/deformable_detr_layers.py
新文件：   mmpose/models/heads/transformer_heads/transformers/detr_layers.py
新文件：   mmpose/models/heads/transformer_heads/transformers/utils.py
修改：     mmpose/models/necks/__init__.py
新文件：   mmpose/models/necks/chanel_mapper_neck.py
修改：     tests/test_datasets/test_transforms/test_bottomup_transforms.py

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/codecs/edpose_label.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        keypoints_res = np.zeros_like(keypoints)
+        keypoints_res[..., 0::3] = Z_pred[..., 0::2]
+        keypoints_res[..., 1::3] = Z_pred[..., 1::2]
+        keypoints_res[..., 2::3] = V_pred[..., 0::1]
+
+        keypoint = keypoints_res.reshape(-1, num_body_points, 3)[:, :, :2]
+
+        return keypoint, scores, boxes
+
+    def box_xyxy_to_cxcywh(self, x):
```
Comment (Ben-Louis) [inline mmpose/codecs/edpose_label.py:167]: Maybe we can use `bbox_xyxy2cs` to replace this method

File: mmpose/codecs/edpose_label.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    def encode(
+        self,
+        img_shape,
+        keypoints: np.ndarray,
+        keypoints_visible: Optional[np.ndarray] = None,
+        area: Optional[np.ndarray] = None,
+        bboxes: Optional[np.ndarray] = None,
+    ) -> dict:
+        """Encoding keypoints 、area、bbox from input image space to normalized
```
Comment (Ben-Louis) [inline mmpose/codecs/edpose_label.py:57]:
```suggestion
        """Encoding keypoints, area, and bbox from input image space to normalized
```

File: mmpose/version.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # Copyright (c) Open-MMLab. All rights reserved.
 
-__version__ = '1.1.0'
+__version__ = '1.0.0'
```
Comment (Tau-J) [inline mmpose/version.py:3]: Please remove this modification

File: mmpose/models/data_preprocessors/data_preprocessor.py [hunk 4]
```diff
@@ -97,3 +157,92 @@ def _get_pad_shape(self, data: dict) -> List[tuple]:
...
+                        assert _batch_input.dim(
+                        ) == 3 and _batch_input.shape[0] == 3, (
+                            'If the mean has 3 values, the input tensor '
+                            'should in shape of (3, H, W), but got the tensor '
+                            f'with shape {_batch_input.shape}')
+                    if self.normalize_bakend == 'cv2':
+                        _batch_input = (_batch_input - self.mean) / self.std
+                    elif self.normalize_bakend == 'pillow':
+                        _batch_input = tvF.normalize(
+                            _batch_input, mean=self.mean, std=self.std)
```
Comment (Tau-J) [inline mmpose/models/data_preprocessors/data_preprocessor.py:198]: `normalize_backend` seems to be confusing, since the normalization is actually conducted by `torchvision`.
If we want to emphasize the img is loaded by `pillow`, maybe `imdecode_backend` is more explicit, just like mmcv does https://github.com/open-mmlab/mmcv/blob/b361a81ad4c5d31611dad7e8d5fbcb7d7df22f34/mmcv/transforms/loading.py#L33

File: mmpose/models/heads/transformer_heads/edpose_head.py [hunk 5]
```diff
@@ -0,0 +1,1349 @@
...
+                    dim=0)
+
+            reference_points = new_reference_points.detach()
+            intermediate_reference_points.append(reference_points)
+
+        return [[itm_out.transpose(0, 1) for itm_out in intermediate],
+                [
+                    itm_refpoint.transpose(0, 1)
+                    for itm_refpoint in intermediate_reference_points
+                ]]
```
Comment (Tau-J) [inline mmpose/models/heads/transformer_heads/edpose_head.py:287]: Please refine this code

File: mmpose/models/heads/transformer_heads/edpose_head.py [hunk 6]
```diff
@@ -0,0 +1,1349 @@
...
+                _out_bbox.shape[1] == _out_keypoint.shape[1]
+
+        out = {
+            'pred_logits': outputs_class[-1],
+            'pred_boxes': outputs_coord_list[-1],
+            'pred_keypoints': outputs_keypoints_list[-1]
+        }
+
+        return out
+
```
Comment (Tau-J) [inline mmpose/models/heads/transformer_heads/edpose_head.py:588]: It seems not to be necessary to pack outputs into dict

File: mmpose/models/heads/transformer_heads/base_transformer_head.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        out_head (ConfigDict, optional): Config for the
+            bounding final out head module. Defaults to None.
+        positional_encoding (ConfigDict, optional): Config for
+            transformer position encoding. Defaults to None.
+        num_queries (int): Number of query in Transformer.
+        loss (ConfigDict, optional): Config for loss functions.
+            Defaults to None.
+        init_cfg (ConfigDict, optional): Config to control the initialization.
+    """
+    _version = 2
```
Comment (Tau-J) [inline mmpose/models/heads/transformer_heads/base_transformer_head.py:36]:
```suggestion
```

**Review by Tau-J — APPROVED** [review]
LGTM
